service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java

```diff
@@ -0,0 +1,106 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@DirtiesContext
+@SpringBootTest(classes = {LeonardoConfig.class})
+@TestPropertySource(properties = {"twds.instance.workspace-id=90e1b179-9f83-4a6f-a8c2-db083df4cd03"})
+class LeonardoDaoTest {
+    @Autowired
+    LeonardoDao leonardoDao;
+
+    @MockBean
+    LeonardoClientFactory leonardoClientFactory;
+
+    final AppsV2Api mockAppsApi = mock(AppsV2Api.class);
+
+    @BeforeEach
+    void beforeEach() {
+        given(leonardoClientFactory.getAppsV2Api(any())).willReturn(mockAppsApi);
+    }
+
+    final String expectedUrl = "https://lz28e3b6da2a7f727cae07c307f6c7c11d96da0d1fde444216.servicebus.windows.net/wds-e433ab84-1725-4666-b07e-4ee7ca";
+    @Test
+    void testWdsUrlNotReturned() throws ApiException {
+        final int statusCode = HttpStatus.INTERNAL_SERVER_ERROR.value();
+        given(mockAppsApi.listAppsV2(String.valueOf(UUID.randomUUID()), "things", null, null))
+                .willThrow(new org.broadinstitute.dsde.workbench.client.leonardo.ApiException(statusCode, "Intentional error thrown for unit test"));
+        var exception = assertThrows(LeonardoServiceException.class, () -> leonardoDao.getWdsEndpointUrl(any()));
+        assertEquals(statusCode, exception.getRawStatusCode());
+    }
+
+    @Test
+    void testWdsUrlReturned() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.RUNNING,1));
+        assertEquals(expectedUrl, url);
+    }
+
+    @Test
+    void testWdsUrlNotFoundWrongStatus() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.DELETED, 1));
+        assertEquals(null, url);
+    }
+
+    @Test
+    void testWdsUrlNotFoundWrongKey() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("not-wds", AppStatus.RUNNING, 1));
+        assertEquals(null, url);
+    }
+
+    @Test
+    void testWdsUrlMultiple() throws ApiException {
+        // tests the case if there are 2 running wds apps
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.RUNNING,2));
+        assertEquals(expectedUrl, url);
+    }
+
+    String buildAppResponseAndCallExtraction(List<ListAppResponse> responses ){
+        var url = leonardoDao.extractWdsUrl(responses);
+        return url;
+    }
+
+    List<ListAppResponse> generateListAppResponse(String wdsKey, AppStatus wdsStatus, int count) {
+        List<ListAppResponse> responseList = new ArrayList<>();
+        var url = expectedUrl;
+        while(count != 0) {
+            ListAppResponse response = mock(ListAppResponse.class);
+            Map<String, String> proxyUrls = new HashMap<String, String>();
```

Comment:
from IntelliJ, let Java infer the HashMap types:
```suggestion
            Map<String, String> proxyUrls = new HashMap<>();
```